psydac/cad/geometry.py

```diff
@@ -475,7 +486,87 @@ def refine_nurbs(nrb, ncells=None, degree=None):
                 raise ValueError('The degree {} must be >= {}'.format(degree, nrb.degree))
             nrb.elevate(axis, times=d)
 
+    for axis in range(nrb.dim):
+        decimals = abs(np.floor(np.log10(np.abs(tol))).astype(int))
+        knots, counts = np.unique(nrb.knots[axis].round(decimals=decimals), return_counts=True)
+        counts = multiplicity[axis] - counts
+        counts[counts<0] = 0
+        knots = np.repeat(knots, counts)
+        nrb = nrb.refine(axis, knots)
     return nrb
+
+def refine_knots(knots, ncells, degree, multiplicity=None, tol=1e-9):
+    """
+    This function refines the knot sequence.
+    It contructs a new grid based on the new number of cells, and it adds the new break points to the nrb grid,
+    such that the total number of cells is equal to the new number of cells.
+    We use knot insertion to construct the new knot sequence , so the geometry is identical to the previous one.
+    It also elevates the degree of the nrb object based on the new degree.
```

Comment:
The `refine_knots` function looks very useful indeed!

But in order to avoid code duplication, I think that function `refine_nurbs` should call this function now.

Reply:
they are not identical, one takes optinal argument the other not, `refine_nurbs` will be used in `caid` in https://github.com/pyccel/caid/blob/5bc5428007c8642412762f80c36e8531e56cd15e/tests/test_cad_geometry.py#L8, that's why I kept them seperate.

Reply:
OK